src/runtimes/node/parser/exports.ts

```diff
@@ -8,21 +8,37 @@ import type {
 } from '@babel/types'
 
 import type { ISCExport } from '../in_source_config/index.js'
+import { ModuleFormat, MODULE_FORMAT } from '../utils/module_format.js'
 
 import type { BindingMethod } from './bindings.js'
 import { isModuleExports } from './helpers.js'
 
-// Finds and returns the following types of exports in an AST:
-// 1. Named `handler` function exports
-// 2. Default function export
-// 3. Named `config` object export
-export const getExports = (nodes: Statement[], getAllBindings: BindingMethod) => {
+/**
+ * Traverses a list of nodes and returns:
+ *
+ * 1. Named `config` object export
+ * 2. Default function export
+ * 3. Named `handler` function exports
+ * 4. The module format syntax used in the file: if any `import` or `export`
+ *    declarations are found, this is ESM; if not, this is CJS
+ */
+export const traverseNodes = (nodes: Statement[], getAllBindings: BindingMethod) => {
   const handlerExports: ISCExport[] = []
 
   let configExport: Record<string, unknown> = {}
   let defaultExport: ExportDefaultDeclaration | undefined
+  let inputModuleFormat: ModuleFormat = MODULE_FORMAT.COMMONJS
 
   nodes.forEach((node) => {
+    if (
+      node.type === 'ImportDeclaration' ||
+      node.type === 'ExportNamedDeclaration' ||
+      node.type === 'ExportDefaultDeclaration' ||
+      node.type === 'ExportAllDeclaration'
```

Comment:
oh, this looks a little bit brittle to me is this the AST from esbuild? – according to the naming I think it is an acorn AST. Are there enums for the values?

But can we define the inputModule format to ESM based on that? Whats about typescript or is this already the transpiled code we are parsing here?

Reply:
> oh, this looks a little bit brittle to me is this the AST from esbuild? – according to the naming I think it is an acorn AST. Are there enums for the values?

esbuild doesn't expose an AST. This is from Babel, as you can see here: https://github.com/netlify/zip-it-and-ship-it/blob/d98dd20825b80d78c6ff0e5c4f2b51c7117d6af5/src/runtimes/node/parser/index.ts#L4.

Can you elaborate on why you think this is brittle? The nodes are fully typed, so if there's a mismatch in this or in a future version of the AST lexicon, we'll know?

> But can we define the inputModule format to ESM based on that? Whats about typescript or is this already the transpiled code we are parsing here?

As I explained in the PR body, we're not trying to infer the module type that we'll emit. We're simply trying to see what kind of syntax was used when authoring the function. If we see an `import` or an `export`, we know it's ESM syntax.

For TypeScript functions this won't matter, but we can use this information to detect when people are writing a `.js` function with ESM without a `type: "module"`. That's the main purpose of surfacing this information.